rfd/0001-testing-guidelines.md

```diff
@@ -0,0 +1,197 @@
+- limited deep-diffing of values
+- largely unmaintained
+
+At the same time, there are some features we'd like to retain:
+- organization of tests into testsuites
+- shared setup/teardown code between tests
+- asserts for concise validation
+
+Go's builtin `testing` package covers most of our needs, and gets new features
+with every new Go release. We just need to complement it with:
+- asserts
+- flexible value diffing (got vs want)
+
+## Details
+
+Use the `testing` package as a base and complement it with several libraries.
+
+### Test organization
+
+Use a hierarchy of:
+- `_test.go` files
+- `func Test*` functions
+- [`t.Run`](https://golang.org/pkg/testing/#hdr-Subtests_and_Sub_benchmarks)
+  subtests (aka table-driven tests)
+
+Shared setup/teardown:
+- [`func TestMain`](https://golang.org/pkg/testing/#hdr-Main) for package level
+- a single `func Test*` with subtests for smaller test groups
+- `func setupX` called from multiple `func Test*` for larger test groups
+
+For performance testing use
+[benchmarks](https://golang.org/pkg/testing/#hdr-Benchmarks).
+
+### Parallelization
+
+Call [`t.Parallel()`](https://golang.org/pkg/testing/#T.Parallel) from all
+tests and subtests that don't need to run serially.
+
+### Asserts
+
+Use
+[testify/assert](https://pkg.go.dev/github.com/stretchr/testify/assert?tab=doc)
+or a plain `if` condition, whichever is easiest.
+
+Skim through the `testify/assert` docs, it has many convenient helpers.
+
+### Diffing
+
+For non-trivial comparisons (such as deeply-nested structs or protobufs), use
+[go-cmp](https://pkg.go.dev/github.com/google/go-cmp/cmp?tab=doc) with
+[cmpopts](https://pkg.go.dev/github.com/google/go-cmp@v0.5.1/cmp/cmpopts?tab=doc).
+
+It allows you to customize checking by ignoring types/fields, equating empty
+and nil slices, approximating float comparisons. See examples below.
+
+### Wishlist
+
+Some things we don't have a good solution for yet:
+- log output collection
+  - if something is logged by a function under test and that test fails, print
+    the log output
+  - it's currently hard to separate logging from a function under test and
+    unrelated logs by a background goroutine
+- performance regression tracking
+  - `go test` will report runtimes of individual tests/subtests
+  - but we don't have tooling to track those over time
+  - note: [benchmarks](https://golang.org/pkg/testing/#hdr-Benchmarks) are an
+    alternative, but must be written manually
+
+## Migration
+
+Re-writing all of teleport tests is a large task. To make it feasible,
+migration will happen organically over the course of ~1 year after these
+guidelines are approved.
+
+The approach is:
+- no tests are rewritten upfront
+- any new test should follow these guidelines
+- updates to any existing test (as part of PRs) should be bundled with a
+  migration of that test
+
+If after 1 year we're left with un-migrated tests, we will discuss whether it's
+worth to invest into dedicated rewrite.
```

Comment:
Thoughts: `integration` and `services/suite` are two places that would probably benefit from total migrations at some point.  The `integration` package is getting a bit bloated anyway, and `services/suite` is used across multiple packages, so migrating it in pieces might just be unwieldy.

Reply:
This is a good thought.

Reply:
Good idea, called them out specifically. They do need some good refactoring.